File: backend/tests/test_lobe_state.py
from src.lobe_state import LobeState


# Simple sanity tests for the LobeState FSM


def test_basic_switch_and_cooldown():
    ls = LobeState(current_lobe=0, n_lobes=3)
    # Incoming scores favor lobe 1
    scores = [0.1, 2.0, 0.0]
    # Step should start a transition
    ls.step(scores, dt=0.1)
    assert ls.target_lobe == 1
    # Finish transition by stepping enough time
    ls.step(scores, dt=ls.transition_time)
    assert ls.current_lobe == 1
    # Immediately attempt to switch back should be blocked by cooldown
    scores2 = [5.0, 0.1, 0.0]
    ls.step(scores2, dt=0.1)
    assert ls.target_lobe is None


def test_transient_fast_switch():
    ls = LobeState(
        current_lobe=0, n_lobes=2, transition_time=1.0, transient_threshold=0.5
    )
    scores = [0.0, 3.0]
    # Provide a high transient to accelerate
    ls.step(scores, dt=0.1, transient=1.0)
    # Transition should be in progress
    assert ls.target_lobe == 1
    # High transient reduces effective time by factor -> finish faster
    ls.step(scores, dt=0.25, transient=1.0)
    assert ls.current_lobe == 1

File: runtime-core/tests/test_lobe_state.rs
use runtime_core::lobe_state::LobeState;

#[test]
fn test_basic_switch_and_cooldown() {
    let mut ls = LobeState::new(3);
    let scores = vec![0.1, 2.0, 0.0];
    ls.step(&scores, 0.1, 0.0);
    assert_eq!(ls.target_lobe, Some(1));
    // Finish transition by stepping enough time
    ls.step(&scores, ls.transition_time, 0.0);
    assert_eq!(ls.current_lobe, 1);
    // Attempt quick switch back should be blocked by cooldown
    let scores2 = vec![5.0, 0.1, 0.0];
    ls.step(&scores2, 0.1, 0.0);
    assert_eq!(ls.target_lobe, None);
}

#[test]
fn test_transient_fast_switch() {
    let mut ls = LobeState::new(2);
    ls.transition_time = 1.0;
    ls.transient_threshold = 0.5;
    let scores = vec![0.0, 3.0];
    ls.step(&scores, 0.1, 1.0);
    assert_eq!(ls.target_lobe, Some(1));
    // High transient reduces effective time -> finish faster
    ls.step(&scores, 0.25, 1.0);
    assert_eq!(ls.current_lobe, 1);
}